docs/writing-intel-modules.md

```diff
@@ -0,0 +1,248 @@
+
+
+### Cleanup
+
+We have just added new nodes and relationships to the graph, and we have also updated previously-added ones
+by using `MERGE`.  We now need to delete nodes and relationships that no longer exist, and we do this by simply removing
+all nodes and relationships that have `lastupdated` NOT set to the `update_tag` of this current run.
+
+- Delete all old nodes
+
+    You can see this in our [GCP VPCs example](https://github.com/lyft/cartography/blob/8d60311a10156cd8aa16de7e1fe3e109cc3eca0f/cartography/data/jobs/cleanup/gcp_compute_vpc_cleanup.json#L4).
+    We run `DETACH DELETE` to delete an old node and disconnect it from all other nodes.
+
+ - Delete all old relationships
+
+    You can see this in the GCP VPC example [here](https://github.com/lyft/cartography/blob/8d60311a10156cd8aa16de7e1fe3e109cc3eca0f/cartography/data/jobs/cleanup/gcp_compute_vpc_cleanup.json#L10)
+    and [here](https://github.com/lyft/cartography/blob/8d60311a10156cd8aa16de7e1fe3e109cc3eca0f/cartography/data/jobs/cleanup/gcp_compute_vpc_cleanup.json#L16).
+
+    - Q: We just `DETACH DELETE`'d the node.  Why do we need to delete the relationships too?
+
+    - A: There are cases where the node may continue to exist but the relationships between it and other nodes have changed.
+        Explicitly deleting stale relationships accounts for this case.
+        See this [short discussion](https://github.com/lyft/cartography/pull/124/files#r312277725).
+
+
+
+
+## Error handling principles
+
+- Don't catch the base Exception class when error handling because it makes it difficult to trace.
+
+
+## Schema
+
+- Update the [schema](https://github.com/lyft/cartography/tree/8d60311a10156cd8aa16de7e1fe3e109cc3eca0f/docs/schema)
+with every change!
+
+## Making tests
+
+- Before making tests, read through and follow the setup steps in [the Cartography developer guide](developer-guide.md).
+
+- Add fake data for testing at `tests/data`.  We can see
+the GCP VPC example here: https://github.com/lyft/cartography/blob/0652c2b6dede589e805156925353bffc72da6c2b/tests/data/gcp/compute.py#L2.
+
+- Add unit tests to `tests/unit/cartography/intel`.  See this [example](https://github.com/lyft/cartography/blob/828ed600f2b14adae9d0b78ef82de0acaf24b86a/tests/unit/cartography/intel/gcp/test_compute.py).
+  These tests ensure that `transform*` manipulates the data in expected ways.
+
+- Add integration tests to  `tests/integration/cartography/intel`.  See this [example](https://github.com/lyft/cartography/blob/828ed600f2b14adae9d0b78ef82de0acaf24b86a/tests/integration/cartography/intel/gcp/test_compute.py).
+  These tests assume that you have neo4j running at localhost:7687 with no password, and ensure that nodes loaded to the
+  graph match your mock data.
+
+## Other
+
+- Smaller PRs are much better than larger PRs!  It's so much easier to review smaller chunks of work.
```

Comment:
We prefer and will accept PRs which incrementally add information from a particular data source. Incomplete representations are OK provided they are consistent.

Reply:
Consistent with?

Reply:
Consistent over time. E.g. we don't sync 100% of AWS resources but the resources which exist in the graph don't change across syncs.